osquery/utils/windows/olecf.cpp

```diff
@@ -0,0 +1,554 @@
+/**
+ * Copyright (c) 2014-present, The osquery authors
+ *
+ * This source code is licensed as defined by the LICENSE file found in the
+ * root directory of this source tree.
+ *
+ * SPDX-License-Identifier: (Apache-2.0 OR GPL-2.0-only)
+ */
+
+#include <osquery/logger/logger.h>
+#include <osquery/utils/conversions/tryto.h>
+#include <osquery/utils/conversions/windows/strings.h>
+#include <osquery/utils/conversions/windows/windows_time.h>
+#include <osquery/utils/windows/olecf.h>
+
+#include <boost/algorithm/hex.hpp>
+#include <boost/algorithm/string.hpp>
+
+#include <math.h>
+#include <string>
+#include <string_view>
+#include <vector>
+
+namespace osquery {
+
+// Header is 512 bytes
+struct OlecfHeader {
+  char sig[8];
+  char guid[16];
+  short revision_minor;
+  short revision_major;
+  short byte_order;
+  int sector_size;
+  int short_sector_size;
+  int total_sectors;
+  int sector_sid;
+  int sector_size_stream;
+  int sector_sid_ssat;
+  int total_sectors_ssat;
+  int sector_sid_msat;
+  int total_sectors_msat;
+  char msat[436];
+};
+
+// Directory entries are 128 bytes
+struct OleDirectory {
+  wchar_t name[32];
+  short name_size;
+  char dir_type;
+  char dir_color;
+  int previous_dir;
+  int next_dir;
+  int sub_dir;
+  char class_identifier[16];
+  int user_flags;
+  char creation_time[8];
+  char modified_time[8];
+  int sector_sid;
+  int directory_size;
+  int reserved;
+};
+
+struct OleDestListHeader {
+  int version;
+  int entries;
+  int pinned_entries;
+  int unknown;
+  int last_entry;
+  int unknown2;
+  int revision;
+  int unknown3;
+};
+
+struct OleDestList {
+  char unknown[8];
+  char droid_volume[16];
+  char droid_file[16];
+  char birth_droid_volume[16];
+  char birth_droid_file[16];
+  char hostname[16];
+  int entry_number;
+  int unknown2;
+  int unknown3;
+  char modified_time[8];
+  int pin_status;
+  int unknown4;
+  int interaction_count;
+  char unknown5[8];
+  short path_size;
+};
+
+struct OleDestListFull {
+  OleDestListHeader header;
+  std::vector<OleDestList> list;
+  std::vector<std::string> path;
+};
+
+const unsigned int kOlecfHeaderSize = 512;
+const unsigned int kDirectorySize = 128;
+
+OlecfHeader parseOlecfHeader(const std::vector<char>& header_data) {
```

Comment:
since this PR is now using the `compoundfilreader` library/header this comment should be resolved. Going to resolve all the other comments about the intial OLE CF implementation. Thanks for review it thought